File: bubbler_app/lib/features/graph/graph_feed_ranking.dart
import '../../data/models/graph.dart';
import '../../data/models/post.dart';
import '../../data/models/preferences.dart';
import '../../data/models/topics.dart';

/// Pure ranking / preference annotation helpers shared by
/// [GraphFeedController] — extracted from Swift `GraphFeedViewModel`.
abstract final class GraphFeedRanking {
  static String? normalizedTopicName(String? topic) {
    if (topic == null) return null;
    final normalized = TopicPreferenceList.normalizedTopic(topic);
    return normalized.isEmpty ? null : normalized;
  }

  static bool containsTopic(String? normalizedTopic, List<String> topics) {
    if (normalizedTopic == null) return false;
    final needle = normalizedTopic.toLowerCase();
    return topics.any((topic) => topic.toLowerCase() == needle);
  }

  static GraphFeedNode makeNode(Post post, UserPreferences preferences) {
    final normalizedTopic = normalizedTopicName(post.topic);
    return GraphFeedNode(
      post: post,
      isPreferredTopic:
          containsTopic(normalizedTopic, preferences.preferredTopics),
      isBlacklistedTopic:
          containsTopic(normalizedTopic, preferences.blacklistedTopics),
    );
  }

  static List<Post> uniqued(List<Post> posts) {
    final seen = <String>{};
    return posts.where((post) => seen.add(post.id)).toList();
  }

  /// Annotates posts, then promotes preferred-topic nodes while preserving
  /// relative order among equals (stable sort on original index).
  static List<GraphFeedNode> rankedNodes(
    List<Post> posts,
    UserPreferences preferences,
  ) {
    final nodes = uniqued(posts)
        .map((post) => makeNode(post, preferences))
        .toList();

    final indexed = [
      for (var i = 0; i < nodes.length; i++) (index: i, node: nodes[i]),
    ];

    indexed.sort((lhs, rhs) {
      if (lhs.node.isPreferredTopic != rhs.node.isPreferredTopic) {
        return lhs.node.isPreferredTopic ? -1 : 1;
      }
      return lhs.index.compareTo(rhs.index);
    });

    return [for (final entry in indexed) entry.node];
  }

  /// First usable session node plus remaining non-blacklisted queue entries.
  /// Returns `null` when the proposed current post is missing or blacklisted.
  static List<GraphFeedNode>? usableSessionNodes(List<GraphFeedNode> ranked) {
    if (ranked.isEmpty) return null;

    final proposedCurrent = ranked.first;
    if (proposedCurrent.isBlacklistedTopic) return null;

    final remaining = ranked
        .skip(1)
        .where((node) => !node.isBlacklistedTopic)
        .toList();

    return [proposedCurrent, ...remaining];
  }

  static String statusMessage({
    required GraphFeedNode node,
    required String? seedStrategyLabel,
    required String defaultMessage,
  }) {
    final parts = <String>[];

    if (node.isPreferredTopic) {
      final topicName = node.topicName;
      if (topicName != null) {
        parts.add('Preferred: $topicName');
      }
    }

    if (seedStrategyLabel != null) {
      parts.add(seedStrategyLabel);
    }

    if (parts.isEmpty) return defaultMessage;
    return parts.join(' · ');
  }
}

File: BubblerAndroid/app/src/main/java/com/bubbler/android/features/graph/GraphFeedRanking.kt
package com.bubbler.android.features.graph

import com.bubbler.android.data.model.GraphFeedNode
import com.bubbler.android.data.model.Post
import com.bubbler.android.data.model.TopicPreferenceList
import com.bubbler.android.data.model.UserPreferences

/**
 * Pure ranking / preference annotation helpers shared by [GraphFeedViewModel].
 * Extracted from Swift `GraphFeedViewModel` private helpers.
 */
internal object GraphFeedRanking {
    fun rankedNodes(posts: List<Post>, preferences: UserPreferences): List<GraphFeedNode> {
        val nodes = uniqued(posts).map { makeNode(it, preferences) }
        return nodes
            .mapIndexed { index, node -> index to node }
            .sortedWith { lhs, rhs ->
                val preferredCmp = when {
                    lhs.second.isPreferredTopic != rhs.second.isPreferredTopic ->
                        if (lhs.second.isPreferredTopic) -1 else 1
                    else -> 0
                }
                if (preferredCmp != 0) preferredCmp else lhs.first.compareTo(rhs.first)
            }
            .map { it.second }
    }

    fun makeNode(post: Post, preferences: UserPreferences): GraphFeedNode {
        val normalizedTopic = normalizedTopicName(post.topic)
        return GraphFeedNode(
            post = post,
            isPreferredTopic = contains(normalizedTopic, preferences.preferredTopics),
            isBlacklistedTopic = contains(normalizedTopic, preferences.blacklistedTopics),
        )
    }

    fun normalizedTopicName(topic: String?): String? {
        if (topic == null) return null
        val normalized = TopicPreferenceList.normalizedTopic(topic)
        return normalized.ifEmpty { null }
    }

    fun contains(normalizedTopic: String?, topics: List<String>): Boolean {
        if (normalizedTopic == null) return false
        return topics.any { it.equals(normalizedTopic, ignoreCase = true) }
    }

    fun uniqued(posts: List<Post>): List<Post> {
        val seen = mutableSetOf<String>()
        return posts.filter { seen.add(it.id) }
    }

    fun statusMessage(
        node: GraphFeedNode,
        seedStrategyLabel: String?,
        defaultMessage: String,
    ): String {
        val parts = buildList {
            if (node.isPreferredTopic) {
                node.topicName?.let { add("Preferred: $it") }
            }
            seedStrategyLabel?.let { add(it) }
        }
        return if (parts.isEmpty()) defaultMessage else parts.joinToString(" · ")
    }
}
